File: banking-transaction/src/main/java/com/banking/transaction/adapters/TransactionController.java
package com.banking.transaction.adapters;

import com.banking.transaction.application.TransactionService;
import com.banking.transaction.application.dto.TransactionRequest;
import com.banking.transaction.application.dto.TransactionResponse;
import com.banking.transaction.application.dto.TransactionReportDTO;
import com.banking.transaction.application.dto.TransactionListResponse;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;

@RestController
@RequestMapping("/transaction")
public class TransactionController {
    private final TransactionService transactionService;

    public TransactionController(TransactionService transactionService) {
        this.transactionService = transactionService;
    }

    @PostMapping
    public ResponseEntity<TransactionResponse> create(@RequestBody TransactionRequest request) {
        return ResponseEntity.ok(transactionService.createTransaction(request));
    }

    @GetMapping
    public ResponseEntity<List<TransactionListResponse>> getAll() {
        return ResponseEntity.ok(transactionService.getAllTransactions());
    }

    @GetMapping("/report")
    public ResponseEntity<List<TransactionReportDTO>> getReport(@RequestParam(required = false) String date) {
        return ResponseEntity.ok(transactionService.getTransactionReport(date));
    }

    @PutMapping("/{id}")
    public ResponseEntity<TransactionResponse> update(@PathVariable Long id, @RequestBody TransactionRequest request) {
        return ResponseEntity.ok(transactionService.updateTransaction(id, request));
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        transactionService.deleteTransaction(id);
        return ResponseEntity.noContent().build();
    }
}

File: frontend/banking-system/src/application/controllers/TransactionController.ts
import { Transaction } from '../../domain/entities/Transaction';
import { ITransactionRepository } from '../../domain/repositories/ITransactionRepository';

export class TransactionController {
    constructor(private repository: ITransactionRepository) {}

    async getAllTransactions(): Promise<Transaction[]> {
        return this.repository.getAll();
    }

    async createTransaction(transaction: Transaction): Promise<Transaction> {
        return this.repository.create(transaction);
    }

    async getTransactionReport(date: string): Promise<Transaction[]> {
        return this.repository.getReport(date);
    }

    async getTransactionReportByClientAndDateRange(startDate: string, endDate: string, clientName: string): Promise<any> {
        return this.repository.getReportByClientAndDateRange(startDate, endDate, clientName);
    }

    async updateTransaction(id: number, transaction: Transaction): Promise<Transaction> {
        return this.repository.update(id, transaction);
    }

    async deleteTransaction(id: number): Promise<void> {
        return this.repository.delete(id);
    }
}